integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/ContainerBasedKafkaCluster.java

```diff
@@ -0,0 +1,235 @@
+                        .withEnv("KAFKA_KRAFT_CLUSTER_ID", holder.getKafkaKraftClusterId());
+            }
+            else {
+                return kafkaContainer.dependsOn(this.zookeeper);
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private byte[] propertiesToBytes(Properties properties) {
+        try (var byteArrayOutputStream = new ByteArrayOutputStream()) {
+            properties.store(byteArrayOutputStream, "server.properties");
+            return byteArrayOutputStream.toByteArray();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private Map<String, String> serverPropertiesToBitnamiEnvVar(Properties serverProperties) {
+        Map<String, String> envVars = new HashMap<>();
+        serverProperties.entrySet().forEach(p -> envVars.put("KAFKA_CFG_" + p.getKey().toString().toUpperCase().replace('.', '_'), p.getValue().toString()));
+        return envVars;
+    }
+
+    @Override
+    public String getBootstrapServers() {
+        return brokers.stream()
+                .map(b -> String.format("localhost:%d", b.getMappedPort(KAFKA_PORT)))
+                .collect(Collectors.joining(","));
+    }
+
+    private Stream<GenericContainer<?>> allContainers() {
+        return Stream.concat(
+                this.brokers.stream(),
+                Stream.ofNullable(this.zookeeper));
+    }
+
+    @Override
+    @SneakyThrows
+    public void start() {
+        try {
+            if (zookeeper != null) {
+                zookeeper.start();
+                Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                    Container.ExecResult result = this.zookeeper.execInContainer(
+                            "sh", "-c",
+                            String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s whoami", ZOOKEEPER_PORT));
+                    LOGGER.debug("zookeeper running {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                    return result.getExitCode() == 0;
+                });
+            }
+            Startables.deepStart(brokers.stream()).get(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            stop();
+            throw new RuntimeException("startup failed or timed out", e);
+        }
+
+        if (clusterConfig.isKraftMode()) {
+            Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                Container.ExecResult result = this.brokers.iterator().next().execInContainer(
+                        "sh", "-c",
+                        "kafka-metadata-shell.sh --snapshot /bitnami/kafka/data/__cluster_metadata-0/*.log ls /brokers");
```

Comment:
The metadata-shell is a slow script and a little overkilled. But I think we can improve it in the future.

Reply:
The test harness no longer uses it.